src/kmod/hash-table-generic.c

```diff
@@ -126,67 +219,84 @@ void hashtbl_shutdown_generic(struct HashTbl *hashTblp)
 	list_del(&(hashTblp->genTables));
 	cb_spinunlock(&g_hashtbl_generic_lock);
 
-	hashtbl_clear_generic(hashTblp);
+	// Ensures we are last to iterate through entries
+	__hashtbl_for_each_generic(hashTblp, _hashtbl_delete_callback, NULL);
 
 	HASHTBL_PRINT("hash shutdown inst=%ld alloc=%ld\n",
 		      atomic64_read(&(hashTblp->tableInstance)),
 		      atomic64_read(&(hashTblp->tableAllocs)));
 
-	cb_destroyspinlock(&(hashTblp->tableSpinlock));
+	if (hashTblp->tablePtr) {
+		vfree(hashTblp->tablePtr);
+		hashTblp->tablePtr = NULL;
+	}
+
 	if (hashTblp->hash_cache) {
 		kmem_cache_destroy(hashTblp->hash_cache);
+		hashTblp->hash_cache = NULL;
 	}
-	vfree(hashTblp);
-}
-
-static int _hashtbl_delete_callback(struct HashTbl *	  hashTblp,
-				    struct HashTableNode *nodep, void *priv)
-{
-	return ACTION_DELETE;
+	kfree(hashTblp);
+	hashTblp = NULL;
 }
 
 void hashtbl_clear_generic(struct HashTbl *hashTblp)
 {
-	HASHTBL_PRINT("ADDR=%p TADDR=%p OFFSET=%lu\n", hashTblp,
-		      hashTblp->tablePtr, sizeof(struct HashTbl));
-
 	hashtbl_for_each_generic(hashTblp, _hashtbl_delete_callback, NULL);
 }
 
 void hashtbl_for_each_generic(struct HashTbl *		  hashTblp,
 			      hashtbl_for_each_generic_cb callback, void *priv)
 {
-	int		   i;
-	uint64_t	   numberOfBuckets;
-	struct hlist_head *hashtbl_tbl = NULL;
+	if (!hashTblp) {
+		return;
+	}
+
+	if (atomic64_read(&(hashTblp->tableShutdown)) == 1) {
```

Comment:
Need to handle race conditions around `tableShutdown`.

Reply:
tableShutdown is atomic. The recently created private wrapper to the generic hashtbl iterator should prevent future usage or calls to a specific hashtbl instance.